custom_envs.py

```diff
@@ -0,0 +1,111 @@
+    self.reset()
+
+  def _gen_reward(self):
+    # generate the reward values in the grid (use the numpy array reward_grid)
+    # first, get shortest path to goal for every square in the grid, then take
+    # the inverse of the values to get our P function (the higher it is, the
+    # closer we are to the goal
+    bfs_q = Queue()
+    processed = np.zeros_like(self.reward_grid)
+
+    # append array positions, neighbor's distance to goal to the queue, we will
+    # process a square by setting its proper distance value, adding unprocessed
+    # neighbors, modifying the # processed mask, and popping the current
+    # position from the queue
+    bfs_q.put((self.goal_pos, 0))
+    while not bfs_q.empty():
+      # parse queue item
+      pos, neighbor_dist = bfs_q.get()
+      r, c = pos
+      neighbors = [(r, c+1), (r+1, c), (r, c-1), (r-1, c)]
+
+      # if it is the goal, we don't change the reward value, otherwise, the
+      # shortest path is 1 + neighbor's distance
+      if self.reward_grid[r, c] != 2:
+        self.reward_grid[r, c] = neighbor_dist + 1
+
+      for r_n, c_n in neighbors:
+        # skip positions that are out of bounds
+        if r_n >= self.reward_grid.shape[0]: continue
+        elif c_n >= self.reward_grid.shape[1]: continue
+        # don't process unreachable squares (lava or wall, marked as 1
+        elif self.reward_grid[r_n, c_n] == 1: continue
+        # don't process squares that have already been processed
+        elif processed[r_n, c_n]: continue
+
+        # add the neighbor to the queue with the current square's distance
+        bfs_q.put(((r_n, c_n), self.reward_grid[r, c]))
+
+        # set the current neighbor as processed
+        processed[r_n, c_n] = 1
+
+    # finally, invert the values in the reward grid to get P
+    # use 2 in the numerator so that reaching the goal has reward of 1
+    self.reward_grid = 2 / self.reward_grid
+
+  def _reward(self, last_pos=None):
+    # hacky solution - ideally we raise an exception if last_pos isn't provided
+    # at train time, however, this reward function is called when calling the
+    # step method of the base class, and we cannot pass last_pos to that step
+    # method
+    if self.training and last_pos is not None:
+      # here, we use the spot reward
+      r, c = self.agent_pos
+      r_l, c_l = last_pos
+      # get I_sr (0 if reward has decreased from last timestep)
+      I_sr = int(self.reward_grid[r, c] >= self.reward_grid[r_l, c_l])
+      # get P (reward grid value)
+      P = self.reward_grid[r, c]
+      reward = I_sr * P
+      return reward
+    else:
+      # during evaluation, return the regular reward
+      return super()._reward()
+
+  def step(self, action, last_pos=None):
+    if self.training:
+      if last_pos is None:
+        raise ValueError("Must provide last position of agent to step function \
+          at train time if using progress reward")
+
+    next_state, _, done, _ = super().step(action)
```

Comment:
what are the other blanked out params?

Reply:
One is the reward and the other is an empty dict in all the crossing environments (I think it is a placeholder for different kinds of environments)